### PR TITLE
Add fail action 6, will fallback to serving stale if retry attempts are exhausted

### PR DESCRIPTION
**Pull request by ezelkow1**
Adds a new fail action that will combine actions 2 and 5 so that if retries are exhausted after attempting collapse then it will also check if it can serve stale if it has an object before deciding to go upstream.

This also refactors a tiny bit of the going to origin logic so that when we are in these retry states we can avoid multiple CACHE_LOOKUP hook calls since previously plugins could get hit with these multiple times.  Also fixes some stats issues around counting hit vs. stale or dupe counts.

For now I mainly want to see if this passes all tests since I can't run all locally

### CODE REVIEW

File: tests/gold_tests/cache/cache-read-retry-stale.test.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+Test.Summary = '''
+Test cache_open_write_fail_action = 6 (READ_RETRY_STALE_ON_REVALIDATE) to verify:
+1. READ_RETRY behavior (same as action 5)
+2. Stale fallback when read retries are exhausted
+3. System does not crash under write lock contention with stale objects
+'''
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply tests/gold_tests/cache/cache-read-retry-stale.test.py:26]: yes will address

**Comment by ezelkow1** [comment]
converting to draft, there is one corner case Im trying to work out (who knew trying to defer hook firings until the ultimate result after looping through the SM is known would be complicated :) )

**Comment by ezelkow1** [comment]
[approve ci autest 0]

File: include/proxy/http/HttpTransact.h [hunk 2]
```diff
@@ -703,6 +703,15 @@ class HttpTransact
...
+    /// Set when stale content is served due to cache write lock failure.
+    /// Used to correctly attribute statistics and VIA strings.
+    bool serving_stale_due_to_write_lock = false;
+
+    /// Set when CACHE_LOOKUP_COMPLETE hook is deferred for action 5/6.
+    /// The hook will fire later with the final result once we know if
+    /// stale content will be served or if we're going to origin.
+    bool cache_lookup_complete_deferred = false;
+
     MgmtByte cache_open_write_fail_action = 0;
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply include/proxy/http/HttpTransact.h:717]: I believe the httpSM gets reset and isn't truly re-used across keepalives like that. A new one is allocated which does this reset

File: src/proxy/http/HttpTransact.cc [hunk 3]
```diff
@@ -3270,21 +3363,55 @@ HttpTransact::HandleCacheOpenReadMiss(State *s)
...
-  } else if (s->cache_info.write_lock_state == CacheWriteLock_t::READ_RETRY) {
+  } else if (s->cache_info.write_lock_state == CacheWriteLock_t::READ_RETRY && !s->serving_stale_due_to_write_lock) {
     // We've looped back around due to failing to read during READ_RETRY mode.
     // Don't attempt another cache write - just proxy to origin without caching.
     TxnDbg(dbg_ctl_http_trans, "READ_RETRY cache read failed, bypassing cache");
     s->cache_info.action = CacheAction_t::NO_ACTION;
+  } else if (s->cache_info.write_lock_state == CacheWriteLock_t::READ_RETRY) {
+    // READ_RETRY with serving_stale_due_to_write_lock = true means we found stale content
+    // but can't serve it (e.g., MUST_PROXY auth). Hook already fired, just go to origin.
+    TxnDbg(dbg_ctl_http_trans, "READ_RETRY stale found but requires origin, bypassing cache");
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply src/proxy/http/HttpTransact.cc]: added

File: src/proxy/http/HttpTransact.cc [hunk 4]
```diff
@@ -3207,21 +3233,108 @@ HttpTransact::handle_cache_write_lock(State *s)
...
+        // This is the main benefit of request collapsing: we found a valid cached object.
+        TxnDbg(dbg_ctl_http_trans, "READ_RETRY: found fresh object, serving from cache");
+        s->cache_lookup_result =
+          (freshness == Freshness_t::FRESH) ? CacheLookupResult_t::HIT_FRESH : CacheLookupResult_t::HIT_WARNING;
+
+        if (s->cache_lookup_complete_deferred) {
+          s->cache_lookup_complete_deferred = false;
+          TRANSACT_RETURN(StateMachineAction_t::API_CACHE_LOOKUP_COMPLETE, HandleCacheOpenReadHit);
+        }
+        HandleCacheOpenReadHit(s);
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply src/proxy/http/HttpTransact.cc:3271]: will add

File: src/proxy/http/HttpTransact.cc [hunk 5]
```diff
@@ -2942,14 +2954,26 @@ HttpTransact::HandleCacheOpenReadHit(State *s)
...
-  if (cache_sm.is_readwhilewrite_inprogress())
+  if (cache_sm.is_readwhilewrite_inprogress()) {
     SET_VIA_STRING(VIA_CACHE_RESULT, VIA_IN_CACHE_RWW_HIT);
+  }
+
+  // If serving stale due to write lock failure (actions 2, 3, or 6), adjust VIA to reflect stale serving.
+  // This ensures correct statistics attribution (cache_hit_stale_served instead of cache_hit_fresh).
+  if (s->serving_stale_due_to_write_lock) {
+    TxnDbg(dbg_ctl_http_trans, "Serving stale due to write lock failure, adjusting VIA for statistics");
+    SET_VIA_STRING(VIA_CACHE_RESULT, VIA_IN_CACHE_STALE);
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply src/proxy/http/HttpTransact.cc:2967]: adding

File: src/proxy/http/HttpTransact.cc [hunk 6]
```diff
@@ -3271,20 +3384,50 @@ HttpTransact::HandleCacheOpenReadMiss(State *s)
...
-  ///////////////////////////////////////////////////////////////
+  // If the CACHE_LOOKUP_COMPLETE hook was deferred, decide whether to fire it now.
+  // Only fire when we're at a terminal state for cache lookup:
+  // - NO_ACTION: Not attempting write lock, MISS is the final result
+  // - READ_RETRY: Already retried and still missed, MISS is the final result
+  //
+  // Don't fire yet if action == PREPARE_TO_WRITE and write_lock_state != READ_RETRY,
+  // because we're about to attempt the write lock and may find content on retry.
+  // In that case, the hook will fire later:
+  // - If READ_RETRY finds HIT: HandleCacheOpenReadHitFreshness fires with HIT
```
[the remark this answers is not in the document]
Comment (ezelkow1) [reply src/proxy/http/HttpTransact.cc]: I will add

**Comment by ezelkow1** [comment]
1) This is actually a bug fix, the current code is wrong for ALL write_fail_actions involving stale (including previous 2/3). Before it was actually reporting these as FRESH when they were in fact STALE.

2) Will implement, just instead of always destroying will do it conditionally so we can still get 304's on write lock fails.

3) Just a nice to have, could be done later

4) will just remove the server_error so it uses the default since it did not go upstream and served stale here

5) Sort of agree here too, thinking about just adding a boolean to the what_is_document_freshness instead to denote if it should skip these checks instead of faking it

**Comment by ezelkow1** [comment]
[approve ci]

File: src/proxy/http/HttpTransact.cc [hunk 7]
```diff
@@ -3207,21 +3234,102 @@ HttpTransact::handle_cache_write_lock(State *s)
...
+        if (s->cache_lookup_complete_deferred) {
+          s->cache_lookup_complete_deferred = false;
+          TRANSACT_RETURN(StateMachineAction_t::API_CACHE_LOOKUP_COMPLETE, HandleCacheOpenReadHit);
+        }
+        HandleCacheOpenReadHit(s);
+      } else {
+        // Object is stale. Save it as potential fallback, then trigger actual cache retry.
+        // HandleCacheOpenReadMiss will serve stale fallback (action 6) or go to origin (action 5).
+        if (is_stale_cache_response_returnable(s)) {
+          s->cache_info.stale_fallback = s->cache_info.object_read;
```
Comment (bneradt) [inline src/proxy/http/HttpTransact.cc:3276]: Claude lists this as a potential use after free:

object_read is a non-owning pointer to CacheVC::alternate, memory owned by the CacheVC. When the new CACHE_LOOKUP is triggered, the old CacheVC can be destroyed (e.g., in HttpCacheSM::state_cache_open_read, the old cache_read_vc is overwritten without being explicitly closed in the non-redirect path). At that point, stale_fallback becomes a dangling pointer. Later, in HandleCacheOpenReadMiss, the code does:

```cpp
s->cache_info.object_read = s->cache_info.stale_fallback;
```

This would dereference freed memory. The fix should deep-copy the stale object into owned storage (e.g., HTTPInfo stale_fallback_store as a value member, not a pointer, and copy via CacheHTTPInfo::copy()) before triggering the new lookup.

File: src/proxy/http/HttpTransact.cc [hunk 8]
```diff
@@ -7300,9 +7456,19 @@ HttpTransact::what_is_document_freshness(State *s, HTTPHdr *client_request, HTTP
...
+  // but for write lock failure scenarios we want to serve the stale content directly without
+  // revalidation. The serving_stale_due_to_write_lock flag tracks that we're actually serving
+  // stale content, so VIA strings and statistics can be correctly attributed.
+  // When evaluate_actual_freshness is true, skip this short-circuit to get the real freshness.
+  if (!evaluate_actual_freshness &&
+      (s->cache_open_write_fail_action & static_cast<MgmtByte>(CacheOpenWriteFailAction_t::STALE_ON_REVALIDATE))) {
     if (is_stale_cache_response_returnable(s)) {
-      TxnDbg(dbg_ctl_http_match, "cache_serve_stale_on_write_lock_fail, return FRESH");
+      TxnDbg(dbg_ctl_http_match, "cache_serve_stale_on_write_lock_fail, return FRESH to bypass revalidation");
+      s->serving_stale_due_to_write_lock = true;
```
Comment (bneradt) [inline src/proxy/http/HttpTransact.cc:7471]: Maybe an unexpected modification of state in a function that seems read-only. Perhaps move this out of the function?

**Review by bneradt — APPROVED** [review]

**Comment by cmcfarlen** [comment]
Cherry-picked to 10.2.x
